### PR TITLE
[MXNET-107]Fused GRU implementation for CPU

### PR DESCRIPTION
**Pull request by lihaofd**
## Description ##
In this PR, it aligns with #10104's registration way and creates a fused GRU operator for CPU.
@pengzhao-intel, @TaoLv , @Sherry-zhang 

### New features ###
- Single layer/Multiple layer and unidirectional/bidirectional GRU, including both forward and backward computation.
- Support kAddTo/kNullOp check for both GRU and LSTM

### Unit-test changes ###
- Create new testcase in tests/python/unittests/test_operator.py.
- update testcase in example/rnn/bucketing/cudnn_rnn_bucketing.py
- Check consistency with original GRUCell implementation.

### Performance ###
We have tested performance of sym.RNN and rnn.GRUCell on our local Skylake-6148 with 2 Sockets and 40 cores. Use MKL as blas lib in this performance test.
Test input size is from DS2 default parameters(seq_length = 300, batch_size = 20, input_size = 800, hidden_size = 800).

Layer=1 bidirectional = False

| API | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
| --------           | :-----:                              | :----:                                   |
| rnn.GRUCell  | 0.12                       |  0.277                            |
| this PR            | 0.048                      |  0.106                              |
| speedup         | 2.5x                              |  2.6x                                    | 

Layer=5 bidirectional = True

| API | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
| --------           | :-----:                              | :----:                                   |
| rnn.GRUCell  | 1.275                       |  2.94                            |
| rnn.GRUCell (cuda)  | 1.047                       |  1.653                            |
| rnn.GRUCell (cudnn)  | 0.161                       |  0.442                           |
| this PR            | 0.551                      |  1.72                              |
| speedup -this PR/GRUCell         | 231.4%                              |  171%                                    | 
| speedup -this PR/GRUCell (cuda)        | 190%                              |  96%                                    | 
| speedup -this PR/GRUCell (cudnn)        | 29.2%                              |  25.7%                                   | 

### Convergency Curve ###
We have tested Convergency of FusedGRU on our CPU-Skylake-8180 with 2 Sockets and 56 cores and GPU-P100  by using example/rnn/bucketing/cudnn_rnn_bucketing.py
Test input size is layer = 3, batch_size = 32, num-embed = 800, num-hidden = 800, num-epochs 20
fusedgru_cpu_gpu

## Checklist ##
 - [X] Passed code style checking (make lint).
 - [X] All changes have test coverage.
 - [ ] Code is well-documented.

### CODE REVIEW

**Comment by piiswrong** [comment]
@Sherry-Zhang 
ping
can we get some review on this?

**Comment by yajiedesign** [comment]
look like good

**Comment by TaoLv** [comment]
@szha @piiswrong @pengzhao-intel Finished code rebase. Please take a review. Thanks.

File: src/operator/rnn_impl.h [hunk 1]
```diff
@@ -454,4 +457,803 @@ void LstmBackward(DType* ws,
...
+  const Tensor<cpu, 2, DType> back_bh(back_bh_ptr, Shape2(3, H));
+
+  if (D == UNIDIRECT) {
+    #pragma omp parallel for
+    for (int i = 0; i < N; i++)
+      for (int j = 0; j < H; j++) {
+        y_ptr[i * H + j] = hx[i][j];
+      }
+  } else {
+    #pragma omp parallel for
```
Comment (pengzhao-intel) [inline src/operator/rnn_impl.h]: Get thread number by `mxnet::engine::OpenMP::Get()->GetRecommendedOMPThreadCount()`. 
Please change all omp by this way :)

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 2]
```diff
@@ -185,7 +185,8 @@ def forward(self, inputs, states=None):
             for i in range(self._dir):
                 self.i2h_weight[i].shape = (self._gates*self._hidden_size, inputs.shape[2])
                 self.i2h_weight[i]._finish_deferred_init()
-        if inputs.context.device_type == 'gpu' or self._mode == 'lstm' or self._mode == 'gru':
+        if inputs.context.device_type == 'gpu' or \
+           self._mode == 'lstm' and not (self._dropout and autograd.is_training()):
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: `self._mode in ['lstm', 'gru']`

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 3]
```diff
@@ -186,11 +186,13 @@ def forward(self, inputs, states=None):
                 self.i2h_weight[i].shape = (self._gates*self._hidden_size, inputs.shape[2])
                 self.i2h_weight[i]._finish_deferred_init()
         if inputs.context.device_type == 'gpu' or \
-           self._mode == 'lstm' and not (self._dropout and autograd.is_training()):
+           self._mode in ['lstm', 'gru'] and \
+           not (self._dropout and autograd.is_training()):
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: I just realized that unlike LSTM, GRU inference doesn't have dropout support yet, so this condition needs fixing.

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 4]
```diff
@@ -186,11 +186,13 @@ def forward(self, inputs, states=None):
                 self.i2h_weight[i].shape = (self._gates*self._hidden_size, inputs.shape[2])
                 self.i2h_weight[i]._finish_deferred_init()
         if inputs.context.device_type == 'gpu' or \
-           self._mode == 'lstm' and not (self._dropout and autograd.is_training()):
+           (self._mode == 'lstm' and not (self._dropout and autograd.is_training())) or \
+           (self._mode == 'gru' and not self._dropout):
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: please do a rebase and skip `autograd.is_training` check

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 5]
```diff
@@ -186,13 +186,11 @@ def forward(self, inputs, states=None):
                 self.i2h_weight[i].shape = (self._gates*self._hidden_size, inputs.shape[2])
                 self.i2h_weight[i]._finish_deferred_init()
         if inputs.context.device_type == 'gpu' or \
-           (self._mode == 'lstm' and not (self._dropout and autograd.is_training())) or \
-           (self._mode == 'gru' and not self._dropout):
+           self._mode == 'lstm' and not self._dropout:
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: gru

File: src/operator/rnn_impl.h [hunk 6]
```diff
@@ -454,4 +457,805 @@ void LstmBackward(DType* ws,
...
+    }
+    Tensor<cpu, 2, DType> hx_l = hx[l];
+    Tensor<cpu, 2, DType> x_l(y_tmp, Shape2(T * N, I));
+    GruBackwardSingleLayer<DType>(ws2, tmp_buf, D, T, N, I, H, x_l, hx_l, wx_l, wh_l, y_l, dy_l,
+                                  dhy_l, gateR_l, gateZ_l, gateN_l, Mnh_l, dx_l, dhx_l,
+                                  dwx_l, dwh_l, dbx_l, dbh_l);
+    if (l > 0) {
+      #pragma omp parallel for num_threads(omp_threads)
+      for (int i = 0; i < T * N * D * H; ++i) {
+        dy_l[i] = dx_l[i];
```
Comment (piiswrong) [inline src/operator/rnn_impl.h]: use mem copy instead?

File: src/operator/rnn_impl.h [hunk 7]
```diff
@@ -40,6 +40,9 @@
 #include "./mshadow_op.h"
 #include "./linalg.h"
```
Comment (piiswrong) [inline src/operator/rnn_impl.h]: I missed this last time:
the file doesn't have name space. Please use namespace similar to rnn-inl.h

File: src/operator/rnn_impl.h [hunk 8]
```diff
@@ -40,6 +40,9 @@
...
 
+#define UNIDIRECT 1
```
Comment (piiswrong) [inline src/operator/rnn_impl.h]: Use named enum instead of define

File: src/operator/rnn_impl.h (hunk 7 above)
Comment (lihaofd) [reply src/operator/rnn_impl.h]: fixed, thanks

File: src/operator/rnn_impl.h (hunk 8 above)
Comment (lihaofd) [reply src/operator/rnn_impl.h]: removed define code, thanks!

File: example/rnn/bucketing/cudnn_gru_bucketing.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (szha) [inline example/rnn/bucketing/cudnn_gru_bucketing.py]: don't copy a whole file. rename cudnn_lstm_bucketing.py to cudnn_rnn_bucketing.py and then add a switch for the mode instead.

Comment (lihaofd) [reply example/rnn/bucketing/cudnn_gru_bucketing.py]: fixed, thanks!

**Comment by piiswrong** [comment]
I missed one thing in the previous PRs: looks like backward doesn't support req == kAddTo? Also it looks like req[kParams] isn't even checked

**Comment by piiswrong** [comment]
Please also check for kNullOp and skip when necessary.

**Comment by lihaofd** [comment]
@piiswrong, kNullOp checking is added. Please help to review it. thx

**Comment by piiswrong** [comment]
You need to check for kNullOp for kData and kState too. And when req is kNullOp, nothing should be written to the corresponding output array

File: src/operator/rnn-inl.h [hunk 10]
```diff
@@ -474,6 +495,9 @@ class RNNOp : public Operator{
     CHECK(dw.CheckContiguous());
     CHECK(dhx.CheckContiguous());
     CHECK(dy.CheckContiguous());
+    if (req[rnn_enum::kParams] != kAddTo && req[rnn_enum::kParams] != kNullOp) {
```
Comment (piiswrong) [inline src/operator/rnn-inl.h]: Isn't gradient still going to be overwritten by backward kernel later?

File: src/operator/rnn_impl.h [hunk 11]
```diff
@@ -454,4 +458,796 @@ void LstmBackward(DType* ws,
...
+  DType alpha = 1.0;
+  DType beta = 0.0;
+  const Tensor<cpu, 2, DType> wx(wx_ptr, Shape2(H * 3, I));
+  const Tensor<cpu, 2, DType> wh(wh_ptr, Shape2(H * 3, H));
+  const Tensor<cpu, 2, DType> back_wx(back_wx_ptr, Shape2(H * 3, I));
+  const Tensor<cpu, 2, DType> back_wh(back_wh_ptr, Shape2(H * 3, H));
+  const int omp_threads = mxnet::engine::OpenMP::Get()->GetRecommendedOMPThreadCount();
+  #pragma omp parallel for num_threads(omp_threads)
+  for (int i = 0; i < D * H * 3 * H; ++i) {
+    dwh[i] = 0;
```
Comment (piiswrong) [inline src/operator/rnn_impl.h]: This is still overwriting gradient even when req[kParams] = kAddTo ?

**Comment by ThomasDelteil** [comment]
This seems to have generated flakiness on this test: https://github.com/apache/incubator-mxnet/issues/11202

**Comment by lihaofd** [comment]
@ThomasDelteil , we will look into the issue.
Thanks!

**Pull request merged**